### ISSUE TITLE
[RFC] sos-4.0 Planning and Discussion

### ISSUE THREAD

**Issue by TurboTurtle** [sosreport/sos#1959]
The last major release (sos-3.0) was made just over six years ago in 2013. This release added the general policy framework allowing sos to be ported to multiple distributions, improved report archive handling and numerous improvements to the plugin and reporting frameworks. In the time since 3.0 was released (when sosreport was available on Fedora and Red Hat Enterprise Linux only), sos has been adopted by numerous additional distributions, products and users.
 
In planning a new major release the team hopes to address several long-standing shortcomings and to deliver a more integrated and streamlined set of tools for gathering, exchanging and reporting crucial support information for today's Linux based systems.
 
Our current goals for sos-4.0 address several long standing limitations in the existing tools, and integrate additional functionality that has historically been covered by third party projects and ad-hoc solutions in specific projects and products. These goals include changes to the core sos reporting tools as used today (for e.g. to improve IO efficiency and memory usage), as well as integrating functionality for collecting sos data from multple host systems and the automated removal of a greater range of potentially sensitive or privacy relevant data included in reports.

Below is the current focus and current goals for a sos-4.0 release.

### Python3 only

For some time now, upstream sos has supported both python 2.7 and python 3. Going forward we will be dropping python 2.7 support. This allows the development team to focus on new features and improvements, rather than backwards compatibility. Further, the beginning of this year marked the end of upstream support for Python 2.7 from the Python Project.

When 4.0 changes start hitting the master branch, the latest 3.9 work will be saved to an alternate branch for ease of maintenance for downstreams.

### Major binary redesign

sosreport has been a staple of diagnostic data collection for years now, and we want to expand on that success. With sos-4.0, we will be introducing a new `sos` command to replace the well-known `sosreport` command as the entry point for the sos project.

What this will allow us to do is add new features from related projects (or those frequently asked for over the years) natively into the sos project. For example, sos-collector will be brought into the project as the `collect` (or similar) sub command, so that users would now run `sos collect [...]` to collect sosreports from multiple hosts.

Similarly, `sosreport` will become `sos report`. We will of course, for transition purposes, continue to carry an `sosreport` which will be marked as deprecated, and point the user to the new `sos report` style. The deprecated `sosreport` will be limited in regards to leveraging the new functionality available via the new `sos` binary.

An overview of some of the subcommands the project could add are as follows - though these are not necessarily slated for 4.0, as the major focus will be focusing on laying the groundwork to easily integrate these in future releases:

* `collect` - integration of sos-collector, which collects sosreports from multiple nodes at once - this IS planned for 4.0 inclusion
* `clean` - integration of the `soscleaner` utility - this is DESIRED for 4.0 inclusion, however a serious review of required work to do this is needed.
* `diff` - compare two (or more) sosreports and find the pertinent differences
* `info` - quickly summarize local sos installation's capabilities
* `examine` - parse through collected data and report it in an at-a-glance style. Similar to pysos/xsos.

Note that this redesign will fundamentally reshape the tree, which means any outstanding PRs would need non-trivial rebasing work by their authors.

### Plugin/API Changes

  * Command Iterators
  Over the last few releases we have been improving the plugin API to allow for easier creation and maintenance of plugins by providing built in methods for common collections. We will be continuing with this trend and with 4.0, targeting high-level iterators, currently being pursued in #1508. That PR is however only a starting point, and we will be looking at adding iterators for more objects, E.G. namespaces.

  * PR #1873 - Container Runtime integration
  This will allow for easier and more stable interactions with not only container runtimes, but also components with other plugins that need to be container-aware, such as the OpenStack and OVN plugins.

  * PR #1788 - Extras Plugin / Ad-hoc collections
  This will allow support organizations and administrators to specify additional collections not covered by the locally installed sos packaging while changes are proposed upstream or waiting for downstream releases.

  * IO improvements 
  Currently sos uses simple file-based IO when processing on-disk IO. In particular, when post-processing large data files, the entire file is read into memory. This is inefficient and causes problems when handling large files. There is some support for limiting the volume of data read into memory from external commands, and limiting the size of collected file data which minimizes these problems in common use cases, however this is currently linked to the --log-size and --all-logs options: use of these options may defeat the limiting and cause significant runtime memory consumption.

  * Non-root support decision
  sos has had limited support in the API for non-root runs for years at this point, however it has never really been utilized or requested to be enhanced. The project will need to consider if non-root support should be dropped, or fully embraced and supported.

  * Archive Metadata
  Having hard data on hand to reference with regards to frequently used or unused sos features/options will make maintenance decisions much easier going forward. While the analysis of this data in bulk will rely mostly on downstream Linux vendors, adding this data directly into an archive for ease of reference by both those vendors and upstream would be mutually beneficial. This metadata would be reflective of the sos (report) run itself, E.G. tarball size, options used, execution times, etc.. and not data about the system on which it is running.

### Distributions
There are currently open PRs for COS and Arch Linux support, however those seem to have stalled. The project is looking for contributors to potentially finish that work, and represent those distributions for upstream development.

Of course, the project also welcomes new or existing contributors to expand distribution support elsewhere as well.

### Release Planning
As 3.9 has just released, the proposal is to make 4.0 our next release - in mid-to-late August in keeping with our traditional 6-month release cadence. This would mean there is no further upstream 3.X development, so bug fixes and additional collections for the 3.X line would be up to individual downstream.

This RFC will remain open for 30 days, at which point the project will finalize the goals and decisions and begin development in earnest.

**Comment by battlemidget** [comment]
RFC - Fix and continue updating sos.readthedocs.io

**Comment by TurboTurtle** [comment]
As to the point of non-root support, my vote is for dropping it. An informal canvas of RH support teams show that only a single team (ironically, mine) would have any benefit from it, and at that point the benefit is small. Plus, I believe there is a way to get this non-root data in a sensible way while running sos as root.

**Comment by BryanQuigley** [comment]
**sos.readthedocs.io**  is there an advantage to not just move fully to the Wiki?
**Python3 only** + 1000 (in fact, should we do this first?)
**collect** We have our own outdated juju based setup that does this, very open to merging to one somehow.
**diff/examine** - I was building ScoutPlane for this purpose / trying to make xsos/pysos better to do this.  Is the idea to have each plugin define some other functions for doing diffs/analysis?  I think that's my preference - but open to other ways.
**non-root support** We could set it up to just not care about checking for root access for which plugins to run - many commands and collections would still work fine without root.  Maybe drop the plugin bit, but add a flag to just ignore the root check?
**IO improvements** I think we need more automated benchmarks in general so it's clear if something is an improvement

I'd like to get better automated testing in general - and start to confirm that certain things are really collected - the diff/analysis bit would be ideal to make that much easier to do.

**Comment by TurboTurtle** [comment]
> **Python3 only** + 1000 (in fact, should we do this first?)

I'm imagining that this work would be done in tandem with the tree re-org and binary re-work. There's a handful of compatibility workarounds that can be slowly removed over the course of 4.0 development, but the big thing here in my mind is that we stop putting those workarounds in, and stop testing on python-2.7.

> **collect** We have our own outdated juju based setup that does this, very open to merging to one somehow.

Take a look at [sos-collector cluster definitions](https://github.com/sosreport/sos-collector/blob/master/soscollector/clusters/__init__.py). It was designed around the idea of very easy extension to other projects/products. All a `juju` setup would need for sos-collector would be to define a method of getting a list of node names/addresses some from "master" (or localhost).

For example, for pacemaker we simply run the `pcs cluster status` command, and parse that as it is a well-known format. For oVirt/RHV, we query the manager DB. There are extremely few product-specific bits beyond the "get a list of nodes" code.

From there it's all remote command execution over multiplexed SSH sessions.

> **diff/examine** - I was building ScoutPlane for this purpose / trying to make xsos/pysos better to do this. Is the idea to have each plugin define some other functions for doing diffs/analysis? I think that's my preference - but open to other ways.

What I'm thinking of is `diff` being pointers to differences in command outputs and config files. Basically a user-friendly wrapper around the straight `diff` command. This was originally suggested by @bmr-cymru so I'll let him correct me if I'm understanding that wrong.

As for `examine` one thing I want to make clear is that there is no *analysis* being done by sos here, just summarization. It's all just data formatting/highlighting so that support engineers/reps/etc can get the data they need in a quicker fashion.

My initial thought is to have the summarization in separate plugin-like files, as from what I've experienced the data formatting bits can actually get quite involved on their own, and I don't see a need to pollute the base `Plugin` class with non collection focused methods. E.G. where for report we have "the kernel plugin", for examine we could have "the kernel formatter (plugin?)".

I think it would be easier on us to maintain this way, but I'm all ears for other opinions.

> **non-root support** We could set it up to just not care about checking for root access for which plugins to run - many commands and collections would still work fine without root. Maybe drop the plugin bit, but add a flag to just ignore the root check?

The problem I foresee with that is this scenario:

    end user: I have a problem
    supporting person: Please provide an sosreport
    end user: Here, I ran it as $user
    supporting person: There's no useful data here, please run it again as root
    end user: well why did you say so!?


Granted there are ways to alleviate this, but I feel we need to either fully embrace it (like the current framework envisions) where plugins declare root/non-root ability - we could potentially extend `SoSPredicate` to check rootfulness - or we drop it entirely. If we just "don't care about it", then that leads to users with pain points and expectations that don't necessarily reflect project goals, but at the same time aren't unreasonable given the project's non-committal stance on it.

> **IO improvements** I think we need more automated benchmarks in general so it's clear if something is an improvement
> I'd like to get better automated testing in general - and start to confirm that certain things are really collected - the diff/analysis bit would be ideal to make that much easier to do.

Agreed. I wish Travis (or anything else really) had multiple distro support. We could then define side-by-side runs and compare collected metrics (either from the testing platform or via the porposed metadata collection).

We could probably (almost definitely) do this under Jenkins, but the problem then becomes infrastructure and ongoing maintenance of Jenkins in addition to sos code.

**Comment by TurboTurtle** [comment]
> RFC - Fix and continue updating sos.readthedocs.io

I'm with @BryanQuigley in wondering if it'd be better to fallback on (and overhaul) the wiki, versus readthedocs.

> RFC - Snap and Flatpak support

What exactly do you mean here, I'm not sure I understand? Packaging sos via snaps/flatpaks? Package detection support (wasn't this recently improved for snaps at least?).

**Comment by BryanQuigley** [comment]
> > **collect** We have our own outdated juju based setup that does this, very open to merging to one somehow.
> 
> Take a look at [sos-collector cluster definitions](https://github.com/sosreport/sos-collector/blob/master/soscollector/clusters/__init__.py). It was designed around the idea of very easy extension to other projects/products. All a `juju` setup would need for sos-collector would be to define a method of getting a list of node names/addresses some from "master" (or localhost).
> 
> For example, for pacemaker we simply run the `pcs cluster status` command, and parse that as it is a well-known format. For oVirt/RHV, we query the manager DB. There are extremely few product-specific bits beyond the "get a list of nodes" code.
> 
> From there it's all remote command execution over multiplexed SSH sessions.

The node that has all our relevant info for this is the Juju "jumphost" machine that has all the issue with using runuser on it.  That's what makes it so complicated for us, but I certainly think we need to find a better way forward - it's not a sos limitation, it's a Juju one.

> > **non-root support** We could set it up to just not care about checking for root access for which plugins to run - many commands and collections would still work fine without root. Maybe drop the plugin bit, but add a flag to just ignore the root check?
> 
> The problem I foresee with that is this scenario:
> 
> ```
> end user: I have a problem
> supporting person: Please provide an sosreport
> end user: Here, I ran it as $user
> supporting person: There's no useful data here, please run it again as root
> end user: well why did you say so!?
> ```
> Granted there are ways to alleviate this, but I feel we need to either fully embrace it (like the current framework envisions) where plugins declare root/non-root ability - we could potentially extend `SoSPredicate` to check rootfulness - or we drop it entirely. If we just "don't care about it", then that leads to users with pain points and expectations that don't necessarily reflect project goals, but at the same time aren't unreasonable given the project's non-committal stance on it.

If they have to run it as sosreport --ignore_lack_of_root_access, I really don't expect it to be done accidentally.

**Comment by battlemidget** [comment]
> > RFC - Fix and continue updating sos.readthedocs.io
> 
> I'm with @BryanQuigley in wondering if it'd be better to fallback on (and overhaul) the wiki, versus readthedocs.

Sounds like we want to move from having an automated way of generating documentation to a homegrown way of doing it via the wiki? I'm not sure what value that adds here as whenever you do a new commit that documentation would be generated automatically.

> 
> > RFC - Snap and Flatpak support
> 
> What exactly do you mean here, I'm not sure I understand? Packaging sos via snaps/flatpaks? Package detection support (wasn't this recently improved for snaps at least)?

We need packaging for flatpak/snaps which I dont think exist yet. This should also lower the barrier for getting sosreport on additional distros quicker.

**Comment by bmr-cymru** [comment]
I think the wiki and RTD serve two different roles. There will be a level of overlap but the in-tree documentation (RTD) should describe **what** - i.e. it tells you what APIs there are and describes the classes, methods and functions that make them up, but it's generally pretty light on **how** - that's what we can put in the wiki to provide more guidance in how to do the right thing when working on sos.

You could look at it as a reference manual vs. a HOWTO style.

**Comment by bmr-cymru** [comment]
> I think we need more automated benchmarks in general so it's clear if something is an improvement

We have a bunch of this already since we log runtimes for various operations. I'd like to see this improved and promoted to the kind of "metadata" that @TurboTurtle talks about in the proposal so that it is easier to consume and keep track of.

If we are moving to a model of more asynchronous or streamed IO then we may need to collect additional metrics (i.e. wall time vs. time spent in IO functions), but we can probably address those needs in parallel with the IO changes themselves.

**Comment by TurboTurtle** [comment]
Ok, I see what I was misunderstanding with RTD now, I'm +1 for it being a 4.0 goal.

@battlemidget I'm not familiar with snap or flatpak packaging myself, but I would imagine it's not terribly difficult to add. I'd be +1 for that as well for 4.0.

For the subcommands, I would say this is my view of priority:

1. `collect` - *needs* to be in 4.0. I also think this is the easiest tool to integrate as we generally just need to link it up to the existing `Policy` detection/framework (in place of the *cough*not great*cough* detection done today) and some shared bits of option handling. If I focused exclusively on this integration, I think it is realistic to say it could be done in 2 working weeks.
2. `info` - *needs* to be in 4.0, as this would be the new alternative for `--list-$thing`. Perhaps in later releases, it can be extended to includ detailed per-plugin info as well.
3. `clean` - This would be nice to have in 4.0, but I'm not certain this is feasible given the level of work needed. Most of the obfuscation bits are fine from a first pass review, but the integration as both a separate subcommand and something that can be used via `sos report --clean` makes this integration tougher. I would like to say we start working on this, and if it's ready by August for 4.0 - great. If not - it's priority number 1 for 4.1.
4. `diff` - I think this should target 4.1. By itself I don't see this as necessarily difficult, but the other subcommands take priority in my mind. Of course if one contributor/maintainer wants to make this their focus for 4.0 inclusion, I'm game for that too.
5. `examine` - I would put this as a last place priority, meaning 4.1 initially, with full understanding that it can be punted to 4.2. There's a lot of planning that needs to go into how we format the data, and **which** data we format and highlight to users.

**Comment by pmoravec** [comment]
> As for `examine` one thing I want to make clear is that there is no _analysis_ being done by sos here, just summarization. It's all just data formatting/highlighting so that support engineers/reps/etc can get the data they need in a quicker fashion.
> 
> My initial thought is to have the summarization in separate plugin-like files, as from what I've experienced the data formatting bits can actually get quite involved on their own, and I don't see a need to pollute the base `Plugin` class with non collection focused methods. E.G. where for report we have "the kernel plugin", for examine we could have "the kernel formatter (plugin?)".
> 
> I think it would be easier on us to maintain this way, but I'm all ears for other opinions.

Isnt such a summarization often user-subjective and use-case-specific? I would vote for just most core data to be summarized there, due to those reasons.

Should we have one summarization per each plugin or just one joint per a report?

"kernel plugin" vs. "kernel formatter" - *if* we unify the way to display the summary for any plugin, then I would vote for extending plugins e.g. by option `in_summary=True` in `add_copy_spec` / `add_cmd_output`. But this has also its drawbacks..

But I see both pros and cons here that fall beyond this generic discussion - I will list them in the separate issue once raised.

**Comment by TurboTurtle** [comment]
Continuing on from #1968 - @bmr-cymru brings up an interesting point with regards to the upload credentials being in `sos.conf`.

Should we split the configuration file, or add other config files? In either case, should we move from `/etc/sos.conf` to `/etc/sos/sos.conf`?

I am currently imagining that in `sos.conf`, we would have separate sections for each subcommand. E.G. a `report` section (which could then have a `plugins` subsection), a `cleaner` section, etc.. But would it make sense to instead have `/etc/sos/{report,cleaner,collector}.conf`?

I personally like the unified config file approach, but I do like the idea of putting credentials in a separate `upload.conf` which I think would almost necessitate moving to an `/etc/sos` directory where these files live.

**Comment by bmr-cymru** [comment]
> I think would almost necessitate moving to an /etc/sos directory where these files live.

It doesn't absolutely require it but it's a lot cleaner (and also I think a good thing to hold off for a major release, i.e. now). It has benefits for SELinux policy authors / relabelling and can simplify packaging. It also gives us a beachead to cleanly establish `.d` directories if we ever needed those for drop-in config file packaging.

**Comment by pmoravec** [comment]
Move `/etc/sos.conf` to `/etc/sos/report.conf` - sounds good (`host` plugin to update).

+1 for the `store-pass` in a dedicated file (only!) - see https://github.com/sosreport/sos/issues/1968#issuecomment-592394447

**Comment by TurboTurtle** [comment]
Coming up on the final stretch here.

It seems we're mostly agreed on the major points, though I'd like to get a bit more feedback for dropping or actually properly implementing non-root support.

To reiterate, on my end, I've not been informed of an actual need for it. None of our support teams see a use-case for asking for non-root sosreports. Since that's the case, I don't see the benefit of carrying the functionality (or for that matter extending it to actually be useful to plugins).

**Comment by BryanQuigley** [comment]
I'm +1 to drop root support.

I'd also like to get feedback on defaulting/(or even switching) plugins to be distro independent - not needing to specify RH/U/D on each one.  I'm not sure what value it gives us at this point.  It is also a likely barrier against new distros joining.   Even the obviously distro specific ones like canonical_livepatch.py would just result in a single command check, it would return not installed and not run.

**Comment by pmoravec** [comment]
Interesting idea to distro-independent majority of plugins (some needs diffs like `apache`). I like it. Maybe we can swap the white-list approach (plugin X to run on distros Y and Z only) to a black-list - if even required?

Or do we have some plugins that must be run on some distro only? Isnt the `packages`/`files`/`kmods` enablement condition sufficient?

**Comment by BryanQuigley** [comment]
There are 41 plugins that don't have any enablement condition (grep -riL -e " files = " -e "packages = " -e "commands = " -e "check_enabled" *)

A quick pass shows that 
16 should have a condition 
22 can always run
unpackaged.py - could just add an rpm check or else would require non-trivial work...

**Comment by TurboTurtle** [comment]
Overall I like the idea, but I have some reservations about implementation. @BryanQuigley do you have an initial idea as to how we'd enable them by default, while still allowing disto-specific overrides? E.G. Are you suggesting extending the existing tagging class design, or replacing it with something else?

**Comment by TurboTurtle** [comment]
We've hit the 30 day mark, but let's leave this open til tomorrow to get any last minute feedback.

**Comment by pmoravec** [comment]
One last-minute idea: implement `xsos` to `sos`: https://github.com/sosreport/sos/issues/1985 ?

**Comment by TurboTurtle** [comment]
> One last-minute idea: implement `xsos` to `sos`: #1985 ?

That was already a goal of `examine`:

> examine - parse through collected data and report it in an at-a-glance style. Similar to pysos/xsos.

**Comment by BryanQuigley** [comment]
>Are you suggesting extending the existing tagging class design, or replacing it with something else?

I see three (not mutually exclusive) possible paths forward:
1. Just drop distro specific checks - current activation criteria are usually enough
2. If we do need it as the only activation criteria - let's just make it match files/packages. distro-like = 
3. Reviewing more plugins I guess we might need to make it a function - is_distro_id("ubuntu") as the way to have distro specific bits in plugins.

One related add - distro detection. Thoughts on moving out of policies? (specify names in policy)
Will sos need to support distros that don't use /etc/os-release? 
I don't want to just rely on the one file, but as the default it is nice.  If we move the detection to core out of policies it also let's us flag cases where multiple conflicting files exist (we get users who modify say distro in lsb_release to install software that was meant for RHEL.. and shockingly that causes other issues).

**Comment by TurboTurtle** [comment]
I don't believe dropping distro checks is in our best interest. The design today allows for easy per-distro changes within a plugin, and is not especially burdensome on plugin authors in my opinion. It also allows for quite easy at-a-glance parsing to see how plugin A will run on distro X. The initial addition of distros _can_ be somewhat annoying with all the tagging classes to be used, but again I don't think its necessarily burdensome so much as tedious.

Adding more `is_distro_id()` -like bits is also not the right direction to me, this detection should be as transparent as possible and adding those kinds of checks throughout plugins only muddies the waters.

As for the policies, I don't think moving the detection out of the class that represents/defines how sos behaves on those distributions buys us enough. If an end user is modifying a Linux install in such a way that it generates conflicting reports of "this is distro XYZ", then that's not something that sos can reasonably be expected to handle. Adding fallback measures for distro detection makes sense, sure, but we can't try to anticipate every breaking-change end users could make to their systems - that's on the individual suppotring organizations/personnel to determine how to respond to those (most likely unsupported) scenarios.


All that being said, I could definitely be persuaded otherwise if functional examples could be provided that show a net benefit to sos. So, as we move through 4.0 work (or beyond, nothing says we can't revisit this particular point down the line in 4.x) if you want to propose some changes, feel free to toss them up in a branch/PR - but I am not especially keen on making this a 4.0 focus as currently described.

**Comment by TurboTurtle** [comment]
We've passed the 30 day period at this point. I think we are generally agreed on the major goals and focus for 4.0 work, and will now close out this RFC. Over the next day or so I'll be breaking out the goals into separate GH Issues for tracking purposes and we can begin development in full.

I'll make an effort to merge any remaining PRs that are in good shape, and then we'll branch 3.9 off from master. If it would help with any downstream distros, we can tag it as a `3.9.1` release as well.

A lot of this work will be gated behind the tree re-org and the new binary design. I'll focus on that, and try to have something posted either by the end of this week or the beginning of next week for review so that we don't lag behind too much in the remaining work.

If there are any remaining changes desired to get into 3.9(.1) before we branch later, please submit them and I will be extra attentive to new PRs.

**Comment by slashdd** [comment]
Are we still on track for "in mid-to-late August" for 4.0 ?

**Comment by TurboTurtle** [comment]
Targeting the 17th for closing 4.0, unless there are major concerns or large contributions that are needed.

Right now, I'd like to get #2192, #2189, #2185, #2172, #2161, and #2158 into master before then. Most of those just need a second ack at this point, though a couple need contributor feedback as to if they will be updated from their current state - #2161 and #2158. Those 2 look fine to me and could be included as-is in my opinion, but see the discussions therein for why there might be an update coming in those (or not).
